### PR TITLE
Important bug fixes, docs, IR spectrographs

### PR DESCRIPTION
**Pull request by jhennawi**
This PR fixes bugs that were introduced by the last load_save_specobj PR that were only discovered after careful analysis of the outputs. Not only was coadd2d broken, but echelle object finding had a subtle but catastrophic object indexing bug.  Since develop is basically currently compromised, this should be approved right away. Tests are running right now. 

There are other things in progress in this PR related to IR spectrographs that are being introduced, fixes to 2d coadding, and documentation.  Given parallel development efforts (i.e. load_save_specobjs and my IR spectrograph PR), there was no simple way to fix these bugs without creating lots of conflicts, so this stuff comes along for the ride for now.

### CODE REVIEW

**Comment by jhennawi** [comment]
Dev suite passes all tests. 

Screen Shot 2019-11-28 at 12 20 15

**Comment by jhennawi** [comment]
Tests passed. Cooked copied over. This one is ready to be approved. 

Screen Shot 2019-11-28 at 13 39 59

**Comment by profxj** [comment]
@jhennawi -- Please put in all the docs needed for `develop` to go into `master` in this PR as originally planned.  I'll then review.  Thanks.

**Comment by jhennawi** [comment]
@profxj this PR is urgent because your PR does not pass coadd2d, and it also contains a bunch of bugs which I've fixed in this PR. For example, your copy method was using pointers and not making copies, wreaking havoc. Develop is not currently fully functional, so I recommend you review and approve this PR first. I'm working on the docs for those codes, but the they also now need to be reworked and fixed, because your PR broke most of the  new coadding, fluxing, and telluric routines. People using develop for those purposes (i.e. anyone doing IR reductions) are now dead in the water, because they cannot combine, flux, or telluric correct their data. I should have admittedly guarded those routines with tests so this is certainly partly my fault, but please also coordinate better when making major changes to the data model as you did with specobjs.

File: pypeit/core/pydl.py [hunk 1]
```diff
@@ -652,7 +653,11 @@ def value(self, x, x2=None, action=None, lower=None, upper=None):
         if self.npoly > 1:
             goodcoeff = self.coeff[:, coeffbk]
         else:
-            goodcoeff = self.coeff[coeffbk]
+            try:
```
Comment (kbwestfall) [inline pypeit/core/pydl.py]: remove try block?

File: pypeit/edgetrace.py [hunk 2]
```diff
@@ -4155,8 +4158,11 @@ def convert_to_tslits_dict(self):
         """
         Stop-gap function to construct the old tslits_dict object.
         """
-        if not self.is_synced:
-            msgs.error('Edges must be synced to construct tslits_dict.')
+
+        # JFH This is causing problems and I cannot figure out how to toggle is_sync when creating psuedo masters. Let's
+        # assume if we are converting to a tslits_dict we are always already synced.
+        #if not self.is_synced:
+        #    msgs.error('Edges must be synced to construct tslits_dict.')
```
Comment (kbwestfall) [inline pypeit/edgetrace.py]: `is_synced` is a direct check of the object itself; it's not a boolean that can be toggled.  If this is being tripped, it means that the object isn't synced and the output of this method is likely to be wrong in some way.  There are some annoying corner cases where the `sync` method fails and the result is that it's not synced.  If you point me to data where this was causing you problems, I can help you figure out why.  But this check needs to be in place for the tslits_dict conversion to be trustworthy.

File: pypeit/edgetrace.py [hunk 3]
```diff
@@ -4220,6 +4228,8 @@ def from_tslits_dict(cls, tslits_dict, master_key, master_dir):
         this.spat_msk = np.zeros(this.spat_cen.shape, dtype=this.bitmask.minimum_dtype())
         this.spat_err = np.zeros(this.spat_cen.shape, dtype=float)
         this.spat_img = np.round(this.spat_cen).astype(int)
+        # JFH This does not work.
+        #this.is_synced = True
```
Comment (kbwestfall) [inline pypeit/edgetrace.py]: It's not a toggle.  If the instance created by the `from_tslits_dict` isn't synced, there may be a problem in how it's being setup. If so, I can help figure out why.

File: pypeit/io.py [hunk 4]
```diff
@@ -187,7 +188,13 @@ def parse_hdr_key_group(hdr, prefix='F'):
...
     # Convert from dictionary with integer keys to an appropriately
     # sorted list
-    return [values[i] for i in range(max(values.keys())+1)]
+    # JFH try/except added to deal with cases where no header cards were found with the desired
+    # prefix. I'm not sure returning an empty list is the desired behavior, but None will cause
+    # downstream things to crash.
+    try:
+        return [values[i] for i in range(max(values.keys())+1)]
+    except:
+        return []
```
Comment (kbwestfall) [inline pypeit/io.py]: Would probably be good to add this as an issue.  Or is it one already that you can reference in the comment?

File: pypeit/scripts/show_2dspec.py [hunk 5]
```diff
@@ -40,6 +38,8 @@ def parser(options=None):
     parser.add_argument('--det', default=1, type=int, help='Detector number')
     parser.add_argument('--showmask', default=False, help='Overplot masked pixels',
                         action='store_true')
+    parser.add_argument('--clean', default=False, help="Remove trace plotting",
+                        action = "store_true")
```
Comment (kbwestfall) [inline pypeit/scripts/show_2dspec.py]: Just an opinion, but `clean` feels non-intuitive.  As a user, I'm not sure I would know what you mean by "Remove" trace plotting.  It's just that you're not overplotting the trace data, right?

File: pypeit/scripts/show_2dspec.py [hunk 6]
```diff
@@ -171,32 +172,32 @@ def main(args):
     # TODO: JFH For some reason Ginga crashes when I try to put cuts in here.
     viewer, ch = ginga.show_image(image, chname=chname_skysub, waveimg=waveimg,
                                   bitmask=mask_in) #, cuts=(cut_min, cut_max),wcs_match=True)
-
-    if sobjs is not None:
-        show_trace(sobjs, args.det, viewer, ch)
+    if not args.clean:
+        if sobjs is not None:
```
Comment (kbwestfall) [inline pypeit/scripts/show_2dspec.py]: ` if not args.clean and sobjs is not None:`

File: pypeit/scripts/show_2dspec.py [hunk 7]
```diff
@@ -171,32 +172,32 @@ def main(args):
...
 
     # SKRESIDS
     chname_skyresids = 'sky_resid-det{:s}'.format(sdet)
     image = (sciimg - skymodel) * np.sqrt(ivarmodel) * (mask == 0)  # sky residual map
     viewer, ch = ginga.show_image(image, chname_skyresids, waveimg=waveimg,
                                   cuts=(-5.0, 5.0), bitmask = mask_in) #,wcs_match=True)
-    if sobjs is not None:
-        show_trace(sobjs, args.det, viewer, ch)
+    if not args.clean:
+        if sobjs is not None:
```
Comment (kbwestfall) [inline pypeit/scripts/show_2dspec.py]: `if not args.clean and sobjs is not None:`

File: pypeit/scripts/show_2dspec.py [hunk 8]
```diff
@@ -171,32 +172,32 @@ def main(args):
...
     chname_resids = 'resid-det{:s}'.format(sdet)
     # full model residual map
     image = (sciimg - skymodel - objmodel) * np.sqrt(ivarmodel) * (mask == 0)
     viewer, ch = ginga.show_image(image, chname=chname_resids, waveimg=waveimg,
-                                  cuts = (-5.0, 5.0), bitmask = mask_in) #,wcs_match=True)
-    if sobjs is not None:
-        show_trace(sobjs, args.det, viewer, ch)
+                                  cuts = (-5.0, 5.0), bitmask = mask_in)
+    if not args.clean:
+        if sobjs is not None:
```
Comment (kbwestfall) [inline pypeit/scripts/show_2dspec.py]: `if not args.clean and sobjs is not None:`

File: pypeit/specobj.py [hunk 9]
```diff
@@ -80,16 +80,23 @@
     'VEL_TYPE': dict(otype=str, desc='Type of heliocentric correction (if any)'),
     'VEL_CORR': dict(otype=float, desc='Relativistic velocity correction for wavelengths'),
     #
-    'DET': dict(otype=(int,np.int64), desc='Detector number'),
+    'DET': dict(otype=(int,np.int64,np.int32), desc='Detector number'),
```
Comment (kbwestfall) [inline pypeit/specobj.py]: Use `(int, np.integer)` instead?

E.g.:
```
>>> x = numpy.int64(3)
>>> y = numpy.int32(3)
>>> z = 3
>>> type(z)
<class 'int'>
>>> isinstance(x, numpy.integer)
True
>>> isinstance(y, numpy.integer)
True
>>> isinstance(z, numpy.integer)
False
```

File: pypeit/spectrographs/keck_mosfire.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+
+        # Set the default exposure time ranges for the frame typing
+        par['calibrations']['standardframe']['exprng'] = [None, 20]
+        par['calibrations']['arcframe']['exprng'] = [20, None]
+        par['calibrations']['darkframe']['exprng'] = [20, None]
+        par['scienceframe']['exprng'] = [20, None]
+        return par
+
+
+#    def default_pypeit_par():
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_mosfire.py]: remove?

File: pypeit/spectrographs/keck_nirspec.py [hunk 11]
```diff
@@ -45,44 +45,113 @@ def __init__(self):
         # Uses default primary_hdrext
         # self.sky_file = ?
 
-    @staticmethod
-    def default_pypeit_par():
+
+    def default_pypeit_par(self):
         """
-        Set default parameters for NIRSPEC reductions
+        Set default parameters for Keck/MOSFIRE
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_nirspec.py]: Think this is a typo.

File: pypeit/spectrographs/keck_nirspec.py [hunk 12]
```diff
@@ -45,44 +45,113 @@ def __init__(self):
...
+
+
+        # Set the default exposure time ranges for the frame typing
+        par['calibrations']['standardframe']['exprng'] = [None, 20]
+        par['calibrations']['arcframe']['exprng'] = [20, None]
+        par['calibrations']['darkframe']['exprng'] = [20, None]
+        par['scienceframe']['exprng'] = [20, None]
         return par
 
+    # JFH Replaced with values from MOSFIRE above.
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_nirspec.py]: Not sure what happened here, but can you check that these changes were intentional?

File: pypeit/spectrographs/keck_nirspec.py [hunk 13]
```diff
@@ -244,7 +313,7 @@ def lamps(self, fitstbl, status):
         raise ValueError('No implementation for status = {0}'.format(status))
 
     # TODO: This function is unstable to shape...
-    def bpm(self, filename, det):
+    def bpm(self, filename, det, shape=None):
```
Comment (kbwestfall) [inline pypeit/spectrographs/keck_nirspec.py]: is this correct?

File: pypeit/core/arc.py [hunk 14]
```diff
@@ -1040,11 +1040,11 @@ def fit_arcspec(xarray, yarray, pixt, fitp):
     # Setup the arrays with fit parameters
     sz_p = pixt.size
     sz_a = yarray.size
-    b      = -1.0*np.ones(sz_p, dtype=np.float)
-    ampl   = -1.0*np.ones(sz_p, dtype=np.float)
-    cent   = -1.0*np.ones(sz_p, dtype=np.float)
-    widt   = -1.0*np.ones(sz_p, dtype=np.float)
-    centerr = -1.0*np.ones(sz_p, dtype=np.float)
+    b      = -999.0*np.ones(sz_p, dtype=np.float)
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/arc.py]: Hi @kbwestfall. Thanks.  Run the LRIS 2-d and GNIRS 2-d coadd on this branch and then try to read in the objects with pypeit_show_2dspec. That is where I discovered the problem with is_synced. So put that line of code back in, and that is where the problem arose.

File: pypeit/core/coadd2d.py [hunk 15]
```diff
@@ -568,6 +580,79 @@ def __init__(self, spec2d_files, spectrograph, det=1, offsets=None, weights='aut
...
+    def optimal_weights(self, slitorderid, objid, const_weights=False):
+        """
+        Determine optimal weights for 2d coadds. This script grabs the information from SpecObjs list for the
+        object with specified slitid and objid and passes to coadd.sn_weights to determine the optimal weights for
+        each exposure.
+
+        Args:
+            slitorderid (int):
+               The slit or order id that has the brightest object whose S/N will be used to determine the weight for each frame.
+            objid (np.ndarray): shape = (nexp,)
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/coadd2d.py]: Okay will move the shape to the description.

File: pypeit/core/coadd2d.py [hunk 16]
```diff
@@ -568,6 +580,79 @@ def __init__(self, spec2d_files, spectrograph, det=1, offsets=None, weights='aut
...
+               Array of object indices of the brightest object whose S/N will be used to determine the weight for each frame.
+            const_weights (bool):
+               Use constant weights for coadding the exposures. Default=False
+
+        Returns:
+            rms_sn, weights
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/coadd2d.py]: Okay I'll give it a go.

File: pypeit/scripts/show_2dspec.py (hunk 5 above)
Comment (feigewang) [reply pypeit/scripts/show_2dspec.py]: changed clean to removetrace to make it clearer

File: pypeit/scripts/show_2dspec.py (hunk 6 above)
Comment (feigewang) [reply pypeit/scripts/show_2dspec.py]: Thanks. fixed

File: pypeit/scripts/show_2dspec.py (hunk 8 above)
Comment (feigewang) [reply pypeit/scripts/show_2dspec.py]: fixed

File: pypeit/scripts/show_2dspec.py (hunk 7 above)
Comment (feigewang) [reply pypeit/scripts/show_2dspec.py]: fixed

**Comment by kbwestfall** [comment]
I fixed the problem with the `is_synced` check.  It was a bug in how the `EdgeTraceSet` instance was being instantiated from the `tslits_dict`.  I've confirmed that the call to `pypeit_show_2dspec` now works with GNIRS.

On another note, your run of the dev suite shows 31 tests, but aren't there 33 now?

File: pypeit/core/arc.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/arc.py]: changed.

File: pypeit/core/coadd2d.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/coadd2d.py]: Can you have a look at this now and tell me if it is correct. I just deleted the first line.

File: pypeit/core/extract.py [hunk 17]
```diff
@@ -513,7 +514,10 @@ def fit_profile(image, ivar, waveimg, thismask, spat_img, trace_in, wave, flux,
     c_answer, cmask   = pydl.iterfit(wave[indsp], flux_sm[indsp], invvar = fluxivar_sm[indsp]*bmask2,
                                      kwargs_bspline={'everyn': 30}, kwargs_reject={'groupbadpix':True,'maxrej':1})
     spline_flux, _ = b_answer.value(wave[indsp])
-    cont_flux, _ = c_answer.value(wave[indsp])
+    try:
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/extract.py]: This is an extremely rare bug that I'm still fixing. Leaving it now so I don't forget.

File: pypeit/core/pydl.py (hunk 1 above)
Comment (jhennawi) [reply pypeit/core/pydl.py]: Same point as above. Will clean it up before we merge.

File: pypeit/core/flat.py [hunk 18]
```diff
@@ -472,6 +472,9 @@ def fit_flat(flat, tilts_dict, tslits_dict_in, slit, inmask = None,
     # ToDo Add some code here to treat the edges and places where fits go bad?
     # Set the pixelflat to 1.0 wherever the flat was nonlinear
     pixelflat[flat >= nonlinear_counts] = 1.0
+    # Do not apply pixelflat field corrections that are greater than 100% to avoid creating edge effects, etc.
+    # TODO Should we do the same for the illumflat??
+    pixelflat = np.fmax(np.fmin(pixelflat, 2.0), 0.5)
```
[the remark this answers is not in the document]
Comment (jhennawi) [reply pypeit/core/flat.py]: Changed.

File: pypeit/specobj.py (hunk 9 above)
Comment (jhennawi) [reply pypeit/specobj.py]: Is there an analog for this for float 32 and float 64. np.float does not work??

File: pypeit/spectrographs/keck_mosfire.py (hunk 10 above)
Comment (jhennawi) [reply pypeit/spectrographs/keck_mosfire.py]: Done.

File: pypeit/spectrographs/keck_nirspec.py (hunk 12 above)
Comment (jhennawi) [reply pypeit/spectrographs/keck_nirspec.py]: Yeah it was, but I'm keeping the old setting sin there for now.

File: pypeit/spectrographs/keck_nirspec.py (hunk 13 above)
Comment (jhennawi) [reply pypeit/spectrographs/keck_nirspec.py]: I think it is now, see also analogous algorithm in spectrograph.

File: pypeit/io.py (hunk 4 above)
Comment (jhennawi) [reply pypeit/io.py]: Issued added.

File: pypeit/spectrographs/keck_nirspec.py (hunk 11 above)
Comment (jhennawi) [reply pypeit/spectrographs/keck_nirspec.py]: fixed.

File: pypeit/core/coadd2d.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypeit/core/coadd2d.py]: No, this still won't get formatted correctly.  Here's what I would write:

```
    Returns:
        tuple: Two numpy.ndarray objects are returned, both have shape
        ``(len(specobjs_list),)``:

            - rms_sn: The root mean square S/N value for each input
              spectrum.
            - weights: The signal-to-noise-squared weights to be applied
              to the spectra.

```

File: pypeit/specobj.py (hunk 9 above)
Comment (kbwestfall) [reply pypeit/specobj.py]: Yeah, `np.floating`.

```
>>> import numpy
>>> x = numpy.float64(2.2)
>>> y = numpy.float32(2.2)
>>> z = 2.2
>>> type(z)
<class 'float'>
>>> isinstance(x, numpy.floating)
True
>>> isinstance(y, numpy.floating)
True
>>> isinstance(z, numpy.floating)
False
```

**Comment by jhennawi** [comment]
Merging

**Pull request merged**